### PR TITLE
[RayJob] add RayJob pass Deadline e2e-test with retry

### PR DESCRIPTION
**Pull request by karta1502545**
## Why are these changes needed?

Add test for rayjob_retry_test.go.

https://github.com/ray-project/kuberay/blob/ea314d7ee780e840382f3de852918f69d7867f25/ray-operator/test/e2e/rayjob_retry_test.go#L83-L87

Make sure RayJob status reaches to JobDeploymentStatusFailed after ActiveDeadlineSeconds regardless of retry times.

## Checks

- [x] I've made sure the tests are passing.
- Testing Strategy
   - [x] Unit tests
   - [ ] Manual tests
   - [ ] This PR is not tested :(

```
$ kind delete cluster

$ sudo rm -rf bin

$ cd ray-operator

$ kind create cluster --image=kindest/node:v1.24.0

$ IMG=kuberay/operator:nightly make docker-build

$ kind load docker-image kuberay/operator:nightly

$ helm install kuberay-operator --set image.repository=kuberay/operator --set image.tag=nightly ../helm-chart/kuberay-operator

$ go test -timeout 30m -v ./test/e2e/rayjob_retry_test.go ./test/e2e/support.go 2>&1 | tee e2e-rayjob-test.log
```
```
--- PASS: TestRayJobRetry (175.48s)
    --- PASS: TestRayJobRetry/Failing_RayJob_without_cluster_shutdown_after_finished (90.70s)
    --- PASS: TestRayJobRetry/Failing_submitter_K8s_Job (77.66s)
    --- PASS: TestRayJobRetry/RayJob_has_passed_ActiveDeadlineSeconds (7.05s)
PASS
ok      command-line-arguments  176.472s
```

### CODE REVIEW

File: ray-operator/test/e2e/rayjob_retry_test.go [hunk 1]
```diff
@@ -127,11 +127,32 @@ func TestRayJobRetry(t *testing.T) {
...
+				WithShutdownAfterJobFinishes(true).
+				WithTTLSecondsAfterFinished(600).
+				WithActiveDeadlineSeconds(5).
+				WithSubmitterPodTemplate(jobSubmitterPodTemplateApplyConfiguration()))
+
+		rayJob, err := test.Client().Ray().RayV1().RayJobs(namespace.Name).Apply(test.Ctx(), rayJobAC, TestApplyOptions)
+		test.Expect(err).NotTo(HaveOccurred())
+		test.T().Logf("Created RayJob %s/%s successfully", rayJob.Namespace, rayJob.Name)
+
+		// The RayJob will transition to `Complete` because it has passed `ActiveDeadlineSeconds`.
```
Comment (andrewsykim) [inline ray-operator/test/e2e/rayjob_retry_test.go:149]: This comment should say `Failed` not `Complete`

File: ray-operator/test/e2e/rayjob_retry_test.go [hunk 2]
```diff
@@ -127,11 +127,32 @@ func TestRayJobRetry(t *testing.T) {
...
-	// TODO: Add a test case to verify that the RayJob has passed ActiveDeadlineSeconds
-	// and ensure that the RayJob transitions to JobDeploymentStatusFailed
-	// regardless of the value of backoffLimit. Refer to rayjob_test.go for an example.
-	// })
+	test.T().Run("RayJob has passed ActiveDeadlineSeconds", func(_ *testing.T) {
+		// RayJob will transition to JobDeploymentStatusFailed
+		// regardless of the value of backoffLimit.
+		rayJobAC := rayv1ac.RayJob("long-running", namespace.Name).
+			WithSpec(rayv1ac.RayJobSpec().
+				WithBackoffLimit(2).
```
Comment (andrewsykim) [inline ray-operator/test/e2e/rayjob_retry_test.go:135]: This test should actually check that the backoffLimit behavior is not used when active deadline passed. You can check this by checking the `status.succeeded` and `status.failed` fields remain at 0

Comment (karta1502545) [reply ray-operator/test/e2e/rayjob_retry_test.go:135]: Why is ```status.failed``` 0? After the job times out, it should be marked as failed. Additionally, when I check the job status, the test output indicates that failed should be 1.
```
test.T().Logf("Waiting for RayJob %s/%s to be 'Failed'", rayJob.Namespace, rayJob.Name)

test.Eventually(RayJob(test, rayJob.Namespace, rayJob.Name), TestTimeoutShort).
	Should(WithTransform(RayJobDeploymentStatus, Equal(rayv1.JobDeploymentStatusFailed)))
test.Expect(GetRayJob(test, rayJob.Namespace, rayJob.Name)).
	To(WithTransform(RayJobReason, Equal(rayv1.DeadlineExceeded)))

test.Expect(GetRayJob(test, rayJob.Namespace, rayJob.Name)).
	Should(WithTransform(RayJobFailed, Equal(int32(0))))
test.Expect(GetRayJob(test, rayJob.Namespace, rayJob.Name)).
	Should(WithTransform(RayJobSucceeded, Equal(int32(0))))
```
```
    rayjob_retry_test.go:157: 
        Expected
            <int32>: 1
        to equal
            <int32>: 0
=== NAME  TestRayJobRetry/RayJob_has_passed_ActiveDeadlineSeconds
    testing.go:1576: test executed panic(nil) or runtime.Goexit: subtest may have called FailNow on a parent test
```

Comment (kevin85421) [reply ray-operator/test/e2e/rayjob_retry_test.go:135]: I think the expected result is `succeeded: 0` and `failed: 1`. When the RayJob passes the deadline, KubeRay will set the `JobDeploymentStatus` to `JobDeploymentStatusFailed`. `status.failed` will increment by 1 if `JobDeploymentStatus` is `JobDeploymentStatusFailed`.

Comment (karta1502545) [reply ray-operator/test/e2e/rayjob_retry_test.go:135]: https://github.com/ray-project/kuberay/blob/cc94c6a71efb74f3c913aed21cbdeb2ede3db30b/ray-operator/controllers/ray/rayjob_controller.go#L383-L385

**Comment by karta1502545** [comment]
Hi, I have added a status check to ensure that `status.failed` is `1` and `status.succeeded` is `0` after a RayJob timeout. 
Thank you very much for your help!

**Review by kevin85421 — APPROVED** [review]
Nice

**Pull request merged**